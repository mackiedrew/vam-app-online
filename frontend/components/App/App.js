import React, { Component } from 'react'

import Tracks from '../Tracks/Tracks'

class App extends Component {

  constructor() {
    super()
  }

  render() {
    return (
      <div className="app">
        <header>
          VAM Online
          <nav>
            <a href="./">Home</a>
          </nav>
        </header>
        <main>
          <Tracks />
          {/* Load Tracks */}
        </main>
        <footer>
          Developed for Valence Audio
        </footer>
      </div>
    )
  }
}

export default App