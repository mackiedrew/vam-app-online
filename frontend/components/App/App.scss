body {
  margin: 0;
}

.app {
  padding: 10px;
}