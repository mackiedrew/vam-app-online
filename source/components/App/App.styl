@require '../../palette.styl'


body
  margin 0
  padding 0
  color $color-text-1
  font-family 'Franklin Gothic Medium', 'Arial Narrow', 'Arial', 'sans-serif'

a
  text-decoration none
  color color-text-1

    
button
  color $color-white
  background-color $color-main
  border 0
  padding 10px
  transition all 0.2s ease

  &:hover
    transition all 0.2s ease
    filter: brightness(1.2)
    cursor pointer

  &:active
    filter: brightness(0.8)
    transition all 0.2s ease

.app
  display flex
  width 100%
  height 100vh
  flex-direction column
  flex-wrap nowrap

  header
    flex-shrink 0
    height 50px
    background-color $color-main
    color white

    h1
      font-size 20px
      display block
      width 100%
      text-align left

  main
    flex-grow 1
    overflow: auto

    .tracks
      width 100%

  footer
    flex-shrink 0