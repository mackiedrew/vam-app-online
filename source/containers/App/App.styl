@require '../../palette.styl'

body
  margin 0
  padding 0
  color $color-text-1
  font-family 'Franklin Gothic Medium', 'Arial Narrow', 'Arial', 'sans-serif'

a
  text-decoration none
  color color-text-1

    
button
  color $color-white
  background-color $color-main
  border 0
  padding 10px
  transition all 0.2s ease

  &:hover
    transition all 0.2s ease
    filter: brightness(1.2)
    cursor pointer

  &:active
    filter: brightness(0.8)
    transition all 0.2s ease

.app

  header
    padding 5px
    background-color $color-main
    color white
    
    h1
      display block
      width 100%
      text-align center

  main
    padding 0

  footer
    width 100%
    text-align center
    font-size 12px
