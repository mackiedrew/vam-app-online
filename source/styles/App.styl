@require './palette.styl'


body
  margin 0
  padding 0
  min-width 300px
  color $color-text-1
  font-family 'Raleway', 'Franklin Gothic Medium', 'Arial Narrow', 'Arial', 'sans-serif'

a
  text-decoration none
  color color-text-1

button
  transition filter 0.2s ease
  border 0
  padding 10px
  background-color $color-main
  color $color-white

  &:hover
    transition filter 0.2s ease
    cursor pointer
    filter brightness(1.2)
    
  &:active
    transition filter 0.2s ease
    filter brightness(0.8)

audio
  display block
  width 500px
  height 100px

.app
  display flex
  width 100%
  height 100vh
  flex-direction column
  flex-wrap nowrap

  main
    display flex
    flex-grow 1
    overflow auto
    flex-wrap nowrap

    .tracks-container
      flex-grow 1

  footer
    flex-shrink 0
