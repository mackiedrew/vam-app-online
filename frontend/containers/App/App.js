import React, { Component } from 'react'
import './App.scss'
import Tracks from '../../components/Tracks/Tracks'

class App extends Component {

  constructor() {
    super()
  }

  render() {
    return (
      <div className="app">
        <header>
          VAM Editor
          <nav>
            <a href="./">Home</a>
          </nav>
        </header>
        <main>
          <Tracks />
          {/* Load Tracks */}
        </main>
        <footer>
          Developed for Valence Audio
        </footer>
      </div>
    )
  }
}

export default App