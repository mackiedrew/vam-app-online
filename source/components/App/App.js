import React, { Component } from "react";
import "./App.styl";

import config from "../../config";

// Libraries
import shortid from "shortid";
import keyboard from "keyboardjs";
import filterObj from "filter-obj";
import { samplesToSeconds } from "../../help/wav/wav";
import { floor } from "../../help/generic/generic";

// Components
import Header from "../../containers/Header/Header";
import AddTrack from "../AddTrack/AddTrack";
import Filters from "../Filters/Filters";
import Tracks from "../Tracks/Tracks";
import Settings from "../Settings/Settings";
import SeekBar from "../SeekBar/SeekBar";

/**
 * This used to be called "Tracks" but was refactored to be the main app since everything was
 * happening here anyway.
 */
class App extends Component {
  constructor(props) {
    // Construct extended class `Component` with passed props
    super(props);

    // Create audio context for editing
    const audioContext = new AudioContext();
    const nextId = shortid.generate();

    // Set initial state to make it easier to reset to later
    this.initialState = {
      playing: false,
      selectedTrack: undefined,
      nextId: nextId,
      filtersOpen: false,
      settingsOpen: false,
      context: audioContext,
      tracks: {},
      trackLengths: {},
      mutedTracks: {},
      seek: 0, // samples
      view: {
        start: 44100 * 60 * 0,
        end: 44100 * 60 * 10
      }
    };
    // Reset state to initialState
    this.state = this.initialState;

    // Bind functions to `this`
    this.seekTo = this.seekTo.bind(this);
    this.simpleAddTracks = this.simpleAddTracks.bind(this);
    this.handleTrackAdd = this.handleTrackAdd.bind(this);
    this.handleTrackRemove = this.handleTrackRemove.bind(this);
    this.reportTrackLength = this.reportTrackLength.bind(this);
    this.toggleFilter = this.toggleFilter.bind(this);
    this.toggleSettings = this.toggleSettings.bind(this);
    this.selectTrack = this.selectTrack.bind(this);
    this.setView = this.setView.bind(this);
    this.viewMagnify = this.viewMagnify.bind(this);
    this.toggleMute = this.toggleMute.bind(this);
    this.togglePlay = this.togglePlay.bind(this);
    this.reportSeek = this.reportSeek.bind(this);
    this.reportPaused = this.reportPaused.bind(this);
    this.play = this.play.bind(this);
    this.pause = this.pause.bind(this);
    this.setTracks = this.setTracks.bind(this);
    this.getAudioTags = this.getAudioTags.bind(this);
    this.componentDidMount = this.componentDidMount.bind(this);
    this.muteTrack = this.muteTrack.bind(this);
    this.selectNextTrack = this.selectNextTrack.bind(this);
    this.selectPreviousTrack = this.selectPreviousTrack.bind(this);
  }

  selectNextTrack() {
    const { tracks, selectedTrack } = this.state;
    const trackIds = Object.keys(tracks);
    const selectedTrackIndex = trackIds.indexOf(selectedTrack);
    const maxTrackIndex = trackIds.length - 1;
    const isLastTrack = selectedTrackIndex >= maxTrackIndex;
    const nextTrackIndex = isLastTrack ? 0 : selectedTrackIndex + 1;
    const nextTrackId = trackIds[nextTrackIndex];
    this.selectTrack(nextTrackId);
  }

  selectPreviousTrack() {
    const { tracks, selectedTrack } = this.state;
    const trackIds = Object.keys(tracks);
    const selectedTrackIndex = trackIds.indexOf(selectedTrack);
    const maxTrackIndex = trackIds.length - 1;
    const isFirstTrack = selectedTrackIndex <= 0;
    const previousTrackIndex = isFirstTrack
      ? maxTrackIndex
      : selectedTrackIndex - 1;
    const previousTrackId = trackIds[previousTrackIndex];
    this.selectTrack(previousTrackId);
  }

  componentDidMount() {
    keyboard.bind([config.play.value, "space"], this.togglePlay);
    keyboard.bind(config.nextTrack.value, this.selectNextTrack);
    keyboard.bind(config.previousTrack.value, this.selectPreviousTrack);
  }

  togglePlay() {
    const { playing } = this.state;
    playing ? this.pause() : this.play();
    this.setState({ playing: !playing });
  }

  getAudioTag(id) {
    const audioTag = document.getElementById(`audio-manager-${id}`);
    return audioTag;
  }

  muteTrack(id, muted) {
    const audioTag = document.getElementById(`audio-manager-${id}`);
    audioTag.mute = muted;
  }

  getAudioTags() {
    const { tracks } = this.state;
    const tracksIds = Object.keys(tracks);
    const audioTags = tracksIds.map(this.getAudioTag);
    return audioTags;
  }

  play() {
    const audioTags = this.getAudioTags();
    this.reportPaused(false);
    audioTags.forEach(tag => tag.play());
  }

  pause() {
    const audioTags = this.getAudioTags();
    this.reportPaused(true);
    audioTags.forEach(tag => tag.pause());
  }

  toggleMute(id) {
    const { mutedTracks } = this.state;
    const muted = id in mutedTracks && mutedTracks[id];
    this.muteTrack(id, !muted);
    const newMutedTracks = {
      ...mutedTracks,
      [id]: !muted
    };
    const newState = { mutedTracks: newMutedTracks };
    this.setState(newState);
  }

  reportSeek(sample) {
    const stateChanges = { seek: sample };
    this.setState(stateChanges);
  }

  reportPaused(paused) {
    const stateChanges = { playing: !paused };
    this.setState(stateChanges);
  }

  setTracks(sample = 0) {
    const audioTags = this.getAudioTags();
    const seconds = samplesToSeconds(sample);
    audioTags.forEach(tag => (tag.currentTime = seconds));
    return seconds;
  }

  enforce() {
    const { seek } = this.state;
    this.setTracks(seek);
  }

  /**
   * Move the seeking cursor to a specified location.
   * @param {Number} sample Integer to move the seeking cursor to, must be integer.
   */
  seekTo(sample = 0) {
    const { trackLengths } = this.state;
    const lengthList = Object.keys(trackLengths).map(key => trackLengths[key]);
    const maxSample = Math.max(...lengthList);
    // Check if sample is less than 0, indexes do not go that low. No upper-bound check yet.
    const candidatePositions = [0, sample, maxSample];
    const sortedPositions = candidatePositions.sort((a, b) => a - b);
    const newPosition = sortedPositions[1];
    this.reportSeek(newPosition);
    this.setTracks(newPosition);
    return newPosition;
  }

  setView({ start, end }) {
    const { trackLengths } = this.state;
    const lengthList = Object.keys(trackLengths).map(key => trackLengths[key]);
    const maxSample = Math.max(...lengthList);

    const endCandidates = [1, end, maxSample * 2];
    const endSorted = endCandidates.sort((a, b) => a - b);
    const newEnd = floor(endSorted[1]);

    const startCandidates = [0, start, newEnd - 1];
    const startSorted = startCandidates.sort((a, b) => a - b);
    const newStart = floor(startSorted[1]);

    const newView = { start: newStart, end: newEnd };
    const stateChanges = { view: newView };
    this.setState(stateChanges);
    return newView;
  }

  viewMagnify(factor) {
    const { view } = this.state;
    const { start, end } = view;

    const newView = {
      start: start,
      end: end * factor
    };
    const actualNewView = this.setView(newView);
    return actualNewView;
  }

  simpleAddTracks(newTracks) {
    const { tracks } = this.state;
    const newTrackList = { ...tracks, ...newTracks };
    const stateChange = { tracks: newTrackList };
    // Set tracks state to be previous state plus new track
    this.setState(stateChange);
  }

  selectTrack(id) {
    this.setState({ selectedTrack: id });
  }

  handleTrackAdd(files) {
    // Add all selected files
    const newIds = files.map(() => shortid.generate());
    const newTracks = newIds.reduce(
      (tracks, id, i) => ({
        ...tracks,
        [id]: files[i]
      }),
      {}
    );

    this.simpleAddTracks(newTracks);

    // Set the current track to the last file added.
    const currentTrackId = newIds.reverse()[0];
    this.selectTrack(currentTrackId);

    // Provide AddTrack with new id after addition.
    const nextId = shortid.generate();
    this.setState({ nextId: nextId });
  }

  toggleFilter() {
    const { filtersOpen } = this.state;
    this.setState({ filtersOpen: !filtersOpen });
  }

  toggleSettings() {
    const { settingsOpen } = this.state;
    this.setState({ settingsOpen: !settingsOpen });
  }

  /**
   * Remove a track from the tracks array matching the provided track id.
   * @param {String} idToRemove ID to remove from the tracks list
   */
  handleTrackRemove(idToRemove = "") {
    // Simple reference to state of tracks
    const { tracks } = this.state;
    // Allow track to remain if it's index does not equal that of the index to remove
    const newTracks = filterObj(tracks, id => id !== idToRemove);
    // Create a 'needs to change' object including all but the specified index
    const stateChange = { tracks: { ...newTracks } };
    // Merge new state object and the old state object
    this.setState(stateChange);
  }

  /**
   * Reporting track length is kind of ugly right now. But basically, the <Track /> will load the 
   * information about their individual files, then report back to the parent. Every time this sort
   * of update is done, it checks to make sure track lengths match existing trackIds. I do this to
   * prevent the state nesting that is frowned upon these days. 
   * @param {String} trackId ID string value from the 'tracks' state list reporting the new length
   * @param {Number} trackLength track length being reported from the trackID
   */
  reportTrackLength(trackId, trackLength) {
    // Breakout 2-layer-deep values for easy reference
    const { tracks, trackLengths } = this.state;
    // Determine the tracks that still exist, then recreate the track lengths array from that
    const stateTrackIds = Object.keys(tracks);
    // Create an object containing existing tracks with their lengths
    const updatedStateTrackLengths = stateTrackIds.reduce(
      (object, id) => ({ ...object, [id]: trackLengths[id] }),
      {}
    );
    // Generate new object containing track lengths that should exist and the new value
    const newTrackLengths = {
      ...updatedStateTrackLengths,
      [trackId]: trackLength
    };
    const stateChange = { trackLengths: newTrackLengths };
    // Change state of track lengths
    this.setState(stateChange);
    // Return for flexibility and testing
    return newTrackLengths;
  }

  render() {
    // Breakout 2-layer-deep values for easy reference
    const {
      nextId,
      seek,
      tracks,
      filtersOpen,
      settingsOpen,
      selectedTrack,
      playing,
      view,
      context,
      mutedTracks
    } = this.state;

    return (
      <div className="app">
        <Header
          toggleFilter={this.toggleFilter}
          toggleSettings={this.toggleSettings}
        >
          <AddTrack handleTrackAdd={this.handleTrackAdd} id={nextId} />
        </Header>
        <main>
          <Filters open={filtersOpen} />
          <Tracks
            context={context}
            handleTrackAdd={this.handleTrackAdd}
            handleTrackRemove={this.handleTrackRemove}
            mutedTracks={mutedTracks}
            playing={playing}
            reportPaused={this.reportPaused}
            reportSeek={this.reportSeek}
            reportTrackLength={this.reportTrackLength}
            seek={seek}
            seekTo={this.seekTo}
            selectTrack={this.selectTrack}
            selectedTrack={selectedTrack}
            toggleMute={this.toggleMute}
            tracks={tracks}
            view={view}
          />
          <Settings open={settingsOpen} />
        </main>

        <footer>
          <SeekBar
            playing={playing}
            seek={seek}
            seekTo={this.seekTo}
            togglePlay={this.togglePlay}
            viewMagnify={this.viewMagnify}
          />
        </footer>

      </div>
    );
  }
}

export default App;
