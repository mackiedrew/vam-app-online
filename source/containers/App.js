// Render
import React, { Component } from "react";
import "../styles/App.styl";

// State
import { connect } from "react-redux";
import { bindActionCreators } from "redux";

// Actions
import generateNextTrackId from "../actions/generateNextTrackId";
import selectedTrackShift from "../actions/selectedTrackShift";
import toggleFiltersMenu from "../actions/toggleFiltersMenu";
import toggleSettingsMenu from "../actions/toggleSettingsMenu";

// Selectors
import longestTrackLength from "../selectors/longestTrackLength";

// Libraries
import keyboard from "keyboardjs";

// Components
import Header from "../components/Header";
import Filters from "../components/Filters";
import Tracks from "../components/Tracks";
import Settings from "../containers/Settings";
import SeekBar from "../containers/SeekBar";

/**
 * This used to be called "Tracks" but was refactored to be the main app since everything was
 * happening here anyway.
 */
class App extends Component {
  constructor(props) {
    super(props);
    props.generateNextTrackId();
    this.componentDidMount = this.componentDidMount.bind(this);
  }

  componentDidMount() {
    const { settings, selectedTrackShift } = this.props;
    const { play, nextTrack, previousTrack } = settings;
    keyboard.bind([play.value, "space"], this.togglePlay);
    keyboard.bind(nextTrack.value, () => selectedTrackShift(1));
    keyboard.bind(previousTrack.value, () => selectedTrackShift(-1));
  }

  render() {
    const {
      filtersOpen,
      trackList,
      toggleSettingsMenu,
      toggleFiltersMenu,
      view
    } = this.props;

    return (
      <div className="app">
        <Header
          toggleFiltersMenu={toggleFiltersMenu}
          toggleSettingsMenu={toggleSettingsMenu}
        />
        <main>
          <Filters open={filtersOpen} />
          <Tracks trackList={trackList} view={view} />
          <Settings />
        </main>
        <footer>
          <SeekBar />
        </footer>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    seekPosition: state.tracks.seekPosition,
    filtersOpen: state.ui.filtersOpen,
    settings: state.settings,
    trackList: state.tracks.trackList,
    view: state.tracks.view
  };
};

const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    {
      toggleFiltersMenu: toggleFiltersMenu,
      toggleSettingsMenu: toggleSettingsMenu,
      generateNextTrackId: generateNextTrackId,
      selectedTrackShift: selectedTrackShift,
      longestTrackLength: longestTrackLength
    },
    dispatch
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(App);
