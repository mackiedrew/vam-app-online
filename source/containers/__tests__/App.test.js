import { App as Subject, mapStateToProps, mapDispatchToProps } from "../App";
jest.mock("../../components/Header", () => "Header");
jest.mock("../../components/Tracks", () => "Tracks");
jest.mock("../Settings", () => "Settings");
jest.mock("../SeekBar", () => "SeekBar");

const mockSettings = {
  grain: {
    label: "Grain Time",
    value: 5,
    unit: "s",
    type: "number"
  },
  quietCutoff: {
    label: "Quietness Threshold",
    value: 10,
    unit: "%",
    type: "number"
  }
};

const mockProps = {
  seekPosition: 10,
  settings: mockSettings,
  trackList: [{ id: "123ABC" }, { id: "456DEF" }],
  view: { start: 0, end: 20 }
};

describe("<App />", () => {
  it("renders without crashing", () => {
    const mockGenerateNextTrackId = sinon.spy();
    const subject = shallow(
      <Subject {...mockProps} generateNextTrackId={mockGenerateNextTrackId} />
    );
    expect(subject.is("div.app")).toBe(true);
    expect(mockGenerateNextTrackId.called).toBe(true);
  });

  it("renders correctly", () => {
    const tree = renderer
      .create(<Subject {...mockProps} generateNextTrackId={sinon.spy()} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("mapStateToProps()", () => {
    const mockState = {
      tracks: {
        seekPosition: 1,
        trackList: 2,
        view: 3
      },
      settings: 4
    };

    it("returns object with expected properties", () => {
      const { seekPosition, settings, trackList, view } = mapStateToProps(
        mockState
      );
      expect(seekPosition).toBe(1);
      expect(trackList).toBe(2);
      expect(view).toBe(3);
      expect(settings).toBe(4);
    });
  });

  describe("mapDispatchToProps()", () => {
    it("returns object with expected properties", () => {
      const mockDispatch = sinon.spy();
      const { toggleSettingsMenu, generateNextTrackId } = mapDispatchToProps(
        mockDispatch
      );
      expect(typeof toggleSettingsMenu).toBe("function");
      expect(typeof generateNextTrackId).toBe("function");
    });
  });
});
