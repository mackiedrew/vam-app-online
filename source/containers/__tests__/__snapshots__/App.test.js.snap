// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<div
  className="app"
>
  <Header
    toggleFiltersMenu={undefined}
    toggleSettingsMenu={undefined}
  />
  <main>
    <Filters
      open={true}
    />
    <Tracks
      trackList={
        Array [
          Object {
            "id": "123ABC",
          },
          Object {
            "id": "456DEF",
          },
        ]
      }
      view={
        Object {
          "end": 20,
          "start": 0,
        }
      }
    />
    <Settings />
  </main>
  <footer>
    <SeekBar />
  </footer>
</div>
`;
